File: org/sathrek/sky_archipelago/worldgen/generator/field/internal/IslandNoise.java
package org.sathrek.sky_archipelago.worldgen.generator.field.internal;

import net.minecraft.util.Mth;

public final class IslandNoise {
   private final long layoutSeed;

   public IslandNoise(long layoutSeed) {
      this.layoutSeed = layoutSeed;
   }

   public int sampleInt(long seed, int salt, int minInclusive, int maxInclusive) {
      if (minInclusive == maxInclusive) {
         return minInclusive;
      }

      double sample = this.sample01(seed, salt);
      return minInclusive + (int)Math.floor(sample * (maxInclusive - minInclusive + 1));
   }

   public int sampleInt(int cellX, int cellZ, int salt, int minInclusive, int maxInclusive) {
      return this.sampleInt(mix(this.layoutSeed + salt, cellX, cellZ), salt, minInclusive, maxInclusive);
   }

   public double sampleRange(long seed, int salt, double minInclusive, double maxInclusive) {
      return Mth.lerp(this.sample01(seed, salt), minInclusive, maxInclusive);
   }

   public double sample01(int cellX, int cellZ, int salt) {
      return this.sample01(mix(this.layoutSeed + salt, cellX, cellZ), salt);
   }

   public double sample01(long seed, int salt) {
      long hash = mix(seed + salt * 31L, salt, 17);
      long masked = hash >>> 11;
      return masked * 1.110223E-16F;
   }

   public double fbm2D(long seed, double x, double z, int octaves, double gain) {
      double value = 0.0;
      double amplitude = 1.0;
      double frequency = 1.0;
      double maxAmplitude = 0.0;

      for (int octave = 0; octave < octaves; octave++) {
         value += this.valueNoise2D(seed + octave * 97L, x * frequency, z * frequency) * amplitude;
         maxAmplitude += amplitude;
         amplitude *= gain;
         frequency *= 2.0;
      }

      return maxAmplitude <= 0.0 ? 0.0 : value / maxAmplitude;
   }

   public double ridgedFbm2D(long seed, double x, double z, int octaves, double gain) {
      return this.toRidged(this.fbm2D(seed, x, z, octaves, gain));
   }

   public double fbm3D(long seed, double x, double y, double z, int octaves, double gain) {
      double value = 0.0;
      double amplitude = 1.0;
      double frequency = 1.0;
      double maxAmplitude = 0.0;

      for (int octave = 0; octave < octaves; octave++) {
         value += this.valueNoise3D(seed + octave * 131L, x * frequency, y * frequency, z * frequency) * amplitude;
         maxAmplitude += amplitude;
         amplitude *= gain;
         frequency *= 2.0;
      }

      return maxAmplitude <= 0.0 ? 0.0 : value / maxAmplitude;
   }

   public double ridgedFbm3D(long seed, double x, double y, double z, int octaves, double gain) {
      return this.toRidged(this.fbm3D(seed, x, y, z, octaves, gain));
   }

   private double valueNoise2D(long seed, double x, double z) {
      int x0 = Mth.floor(x);
      int z0 = Mth.floor(z);
      int x1 = x0 + 1;
      int z1 = z0 + 1;
      double tx = smoothstep(x - x0);
      double tz = smoothstep(z - z0);
      double v00 = this.lattice2D(seed, x0, z0);
      double v10 = this.lattice2D(seed, x1, z0);
      double v01 = this.lattice2D(seed, x0, z1);
      double v11 = this.lattice2D(seed, x1, z1);
      double a = Mth.lerp(tx, v00, v10);
      double b = Mth.lerp(tx, v01, v11);
      return Mth.lerp(tz, a, b);
   }

   private double valueNoise3D(long seed, double x, double y, double z) {
      int x0 = Mth.floor(x);
      int y0 = Mth.floor(y);
      int z0 = Mth.floor(z);
      int x1 = x0 + 1;
      int y1 = y0 + 1;
      int z1 = z0 + 1;
      double tx = smoothstep(x - x0);
      double ty = smoothstep(y - y0);
      double tz = smoothstep(z - z0);
      double c000 = this.lattice3D(seed, x0, y0, z0);
      double c100 = this.lattice3D(seed, x1, y0, z0);
      double c010 = this.lattice3D(seed, x0, y1, z0);
      double c110 = this.lattice3D(seed, x1, y1, z0);
      double c001 = this.lattice3D(seed, x0, y0, z1);
      double c101 = this.lattice3D(seed, x1, y0, z1);
      double c011 = this.lattice3D(seed, x0, y1, z1);
      double c111 = this.lattice3D(seed, x1, y1, z1);
      double x00 = Mth.lerp(tx, c000, c100);
      double x10 = Mth.lerp(tx, c010, c110);
      double x01 = Mth.lerp(tx, c001, c101);
      double x11 = Mth.lerp(tx, c011, c111);
      double y0Mix = Mth.lerp(ty, x00, x10);
      double y1Mix = Mth.lerp(ty, x01, x11);
      return Mth.lerp(tz, y0Mix, y1Mix);
   }

   private double lattice2D(long seed, int x, int z) {
      long hash = mix(seed, x, z);
      return (hash >>> 11) * 1.110223E-16F * 2.0 - 1.0;
   }

   private double lattice3D(long seed, int x, int y, int z) {
      long hash = mix(seed ^ y * -7046029254386353131L, x, z);
      hash ^= Long.rotateLeft(y * -4417276706812531889L, 21);
      hash ^= hash >>> 29;
      hash *= 1609587929392839161L;
      hash ^= hash >>> 28;
      return (hash >>> 11) * 1.110223E-16F * 2.0 - 1.0;
   }

   private double toRidged(double value) {
      return 1.0 - Math.abs(value);
   }

   public static double ellipseDensity(double localX, double localZ, double radiusX, double radiusZ) {
      double xTerm = localX * localX / (radiusX * radiusX);
      double zTerm = localZ * localZ / (radiusZ * radiusZ);
      return 1.0 - (xTerm + zTerm);
   }

   private static double smoothstep(double t) {
      return t * t * (3.0 - 2.0 * t);
   }

   public static long mix(long seed, int x, int z) {
      long hash = seed;
      hash ^= x * -7046029254386353131L;
      hash = Long.rotateLeft(hash, 17);
      hash ^= z * -4417276706812531889L;
      hash ^= hash >>> 29;
      hash *= 1609587929392839161L;
      hash ^= hash >>> 32;
      hash *= -7046029254386353131L;
      return hash ^ hash >>> 28;
   }
}

File: bedrock-addon/scripts/worldgen/IslandNoise.js
export class IslandNoise {
  constructor(layoutSeed) { this.layoutSeed = BigInt.asIntN(64, BigInt(layoutSeed)); }
  static mix(seed, x, z) {
    let h = BigInt.asIntN(64, BigInt(seed));
    h ^= BigInt.asIntN(64, BigInt(x) * -7046029254386353131n);
    h = BigInt.asIntN(64, (h << 17n) | ((h & ((1n << 64n) - 1n)) >> 47n));
    h ^= BigInt.asIntN(64, BigInt(z) * -4417276706812531889n);
    h ^= h >> 29n;
    h = BigInt.asIntN(64, h * 1609587929392839161n);
    h ^= h >> 32n;
    h = BigInt.asIntN(64, h * -7046029254386353131n);
    return BigInt.asIntN(64, h ^ (h >> 28n));
  }
  sample01(seed, salt) {
    const h = IslandNoise.mix(BigInt(seed) + BigInt(salt) * 31n, salt, 17);
    const masked = BigInt.asUintN(64, h) >> 11n;
    return Number(masked) / 9007199254740992;
  }
  sampleInt(seed, salt, min, max) { if (min === max) return min; return min + Math.floor(this.sample01(seed, salt) * (max - min + 1)); }
  sampleCell(cellX, cellZ, salt) { return this.sample01(IslandNoise.mix(this.layoutSeed + BigInt(salt), cellX, cellZ), salt); }
  sampleCellInt(cellX, cellZ, salt, min, max) { return this.sampleInt(IslandNoise.mix(this.layoutSeed + BigInt(salt), cellX, cellZ), salt, min, max); }
  sampleRange(seed, salt, min, max) { return min + (max - min) * this.sample01(seed, salt); }
  lattice2D(seed, x, z) { return this.toSigned(IslandNoise.mix(BigInt(seed), x, z)); }
  lattice3D(seed, x, y, z) {
    let h = IslandNoise.mix(BigInt(seed) ^ BigInt(y) * -7046029254386353131n, x, z);
    h ^= BigInt.asIntN(64, (BigInt(y) * -4417276706812531889n));
    h ^= h >> 29n; h = BigInt.asIntN(64, h * 1609587929392839161n); h ^= h >> 28n;
    return this.toSigned(h);
  }
  toSigned(h) { return (Number(BigInt.asUintN(64, h) >> 11n) / 9007199254740992) * 2 - 1; }
  static smooth(t) { return t * t * (3 - 2 * t); }
  value2D(seed, x, z) {
    const x0 = Math.floor(x), z0 = Math.floor(z), tx = IslandNoise.smooth(x-x0), tz = IslandNoise.smooth(z-z0);
    const a=this.lattice2D(seed,x0,z0),b=this.lattice2D(seed,x0+1,z0),c=this.lattice2D(seed,x0,z0+1),d=this.lattice2D(seed,x0+1,z0+1);
    return (a+(b-a)*tx)+((c+(d-c)*tx)-(a+(b-a)*tx))*tz;
  }
  value3D(seed,x,y,z) {
    const x0=Math.floor(x),y0=Math.floor(y),z0=Math.floor(z),x1=x0+1,y1=y0+1,z1=z0+1;
    const tx=IslandNoise.smooth(x-x0),ty=IslandNoise.smooth(y-y0),tz=IslandNoise.smooth(z-z0);
    const c000=this.lattice3D(seed,x0,y0,z0),c100=this.lattice3D(seed,x1,y0,z0),c010=this.lattice3D(seed,x0,y1,z0),c110=this.lattice3D(seed,x1,y1,z0),c001=this.lattice3D(seed,x0,y0,z1),c101=this.lattice3D(seed,x1,y0,z1),c011=this.lattice3D(seed,x0,y1,z1),c111=this.lattice3D(seed,x1,y1,z1);
    const x00=c000+(c100-c000)*tx,x10=c010+(c110-c010)*tx,x01=c001+(c101-c001)*tx,x11=c011+(c111-c011)*tx;
    return (x00+(x10-x00)*ty)+((x01+(x11-x01)*ty)-(x00+(x10-x00)*ty))*tz;
  }
  fbm2D(seed,x,z,octaves,gain) { let v=0,a=1,f=1,m=0; for(let i=0;i<octaves;i++){v+=this.value2D(BigInt(seed)+BigInt(i)*97n,x*f,z*f)*a;m+=a;a*=gain;f*=2;} return m?v/m:0; }
  ridgedFbm2D(seed,x,z,o,g){return 1-Math.abs(this.fbm2D(seed,x,z,o,g));}
  fbm3D(seed,x,y,z,o,g){let v=0,a=1,f=1,m=0;for(let i=0;i<o;i++){v+=this.value3D(BigInt(seed)+BigInt(i)*131n,x*f,y*f,z*f)*a;m+=a;a*=g;f*=2;}return m?v/m:0;}
  ridgedFbm3D(seed,x,y,z,o,g){return 1-Math.abs(this.fbm3D(seed,x,y,z,o,g));}
  static ellipseDensity(x,z,rx,rz){return 1-(x*x/(rx*rx)+z*z/(rz*rz));}
}
